### ISSUE TITLE
Long RSA keygen time resulting from high loop count in bn_rand_range

### ISSUE THREAD

**Issue by nathanowen42** [openssl/openssl#4151]
Using OpenSSL 1.0.2k

Not sure if this is the correct place to ask about this, if not please redirect me accordingly.

TL;DR: RSA key generation on ARM based SOC generally takes a very long time and that time window has a large standard divination.   Is there a way to prevent this that does not significantly compromise security?  I think there may be a bug with the bn_rand_range function that results in a high loop count.

I recently implemented a proprietary RAND_METHOD that utilizes the TRNG of an Arm based SOC.  On this particular product it was necessary to have the ability to generate cryptographically secure RSA key pairs at runtime.

Once I had optimized my RAND_METHOD and removed my debug statements,  the RSA keygen started failing the majority of the time with following error.

> 3338315896:error:0307A071:bignum routines:BN_rand_range:too many iterations:bn_rand.c:275

After some time I determined that I needed to add a delay between TRNG keystore reads so as to allow it time to sufficiently alter its state.  Otherwise the bn_rand_range function has the tendency to not get the output it wants within the loop limit.  I have not dug into it enough to be sure yet, but I am guessing that bn_rand_range is looping like this to avoid modulo bias or something similar.

Long story short, I was able to play around with the delay to optimize the TRNG based randBytes function such that RSA keygen takes about as long as it does with the default RNG, generally around 50 seconds, but ranging anywhere from ~9 seconds to ~3 minutes with a standard deviation of ~30 seconds.

I find it surprising that key generation has such a large standard time deviation.  My understanding was that de-biasing loops generally were very likely to complete on short order and very unlikely to loop many times.  The comments in the bn_rand_range even say as much.

> If r < 3 x range, use r := r MOD range (which is either r, r -
> range, or r - 2*range). Otherwise, iterate once more. Since
> 3*range = 11..._2, each iteration succeeds with probability >=
> .75.

Normally this would lead me to believe that my TRNG was over-biased towards out-of-range numbers but given that the default random number generator has the same issue, I do not believe this is the case.

My Question is this:
Is there a known way that I can configure/alter OpenSSL such that the RSA key generation completes in a smaller and more consistent time frame without adding significant bias to random number generation?  Is there a bug in the bn_rand_range function resulting in an iteration success rate far lower than the 75% expected?

**Comment by richsalz** [comment]
In a word, no.  Finding primes is a probabilistic effort.

**Comment by kroeckx** [comment]
If you look at any other architecture, you will also see a big difference in the time that it takes to generate it, and I guess that the standard deviation would be about the same.

But you indicate that you're using a TRNG. Have you actually analyzed how much entropy it can provide? Does it somehow get debiased and turned into something that contains around 8 bits of entropy per byte? You're not supposed to directly use the output of the TRNG without processing it somehow first. It seems to me you at least need to take a good like at your TRNG, having to wait seems broken.

**Comment by nathanowen42** [comment]
@richsalz So it is actually trying to find a prime number on the fly...  Fascinating, I had wondered how the RSA primes were determined.  Does OpenSSL actually have to run a full trial division test or does it just use one of the primality tests and hope for the best?

@kroeckx Unfortunately I have not had time to develop functions to test the entropy yet.  Probably just need to check that the probability that a given bit in the 256 bit sequence is a 1 is close to 50%.  I have debated just making the RAND_bytes function call the default random bytes function after calling the default RAND_add with the output of the TRNG.  I think the TRNG is probably based on a ring oscillator but I will need to contact the SOC vendor to know for sure.

**Comment by nathanowen42** [comment]
@richsalz nvm answered my own question https://wiki.openssl.org/index.php/Manual:BN_generate_prime(3).  Evidently it does just use a primality test and hope for the best.  There is a small (2^-80) chance that the number is not actually prime.  I suppose if your luck is that bad you probably have worse things to worry about.
